pkg/assetfetcher/http.go

```diff
@@ -131,50 +103,81 @@ func (c *NexposeAssetFetcher) FetchAssets(ctx context.Context, siteID string, sc
 	return pagedAssetChan, pagedErrChan
 }
 
-func (c *NexposeAssetFetcher) makeRequest(ctx context.Context, wg *sync.WaitGroup, siteID string, scanID string, page int, assetChan chan domain.AssetEvent, errChan chan error) {
-	defer wg.Done()
+// fetchAllNexposeSiteAssets returns all assets from a given Nexpose site. This function first makes a call
+// to retrieve the first page of assets in a Nexpose site, then concurrently retrieves all pages after knowing the total pages.
+// The function uses a fanout pattern with no partial error
+func (c *NexposeAssetFetcher) fetchAllNexposeSiteAssets(ctx context.Context, siteID string) ([]Asset, error) {
+	// have to page through to get all the assets, start with page 0
+	// make the first call to Nexpose to get the total number of pages we'll
+	var currentPage = 0
+	ctx, cancel := context.WithCancel(ctx)
+	firstPageOfAssets, initialErr := c.fetchNexposeSiteAssetsPage(ctx, currentPage, siteID)
+	if initialErr != nil {
+		defer cancel()
+		return []Asset{}, &ErrorFetchingAssets{Inner: initialErr, Page: 0, SiteID: siteID}
+	}
+
+	totalPages := firstPageOfAssets.Page.TotalPages
+	totalAssets := firstPageOfAssets.Resources
+
+	pageAssetChan := make(chan SiteAssetsResponse, totalPages)
+	pageErrChan := make(chan error, totalPages)
+	for currentPage := 1; currentPage < totalPages; currentPage++ {
+		go func(ctx context.Context, page int, site string) {
+			pageOfAssets, err := c.fetchNexposeSiteAssetsPage(ctx, page, site)
+			if err != nil {
+				defer cancel()
+				pageErrChan <- &ErrorFetchingAssets{Inner: err, Page: page, SiteID: siteID}
+				return
+			}
+			pageAssetChan <- pageOfAssets
+		}(ctx, currentPage, siteID)
+	}
+	defer cancel()
```

Comment:
Task: Remove the `cancel` invocation here. I'm pretty sure this is a bug that will cancel all the requests that were just kicked off, if they take longer than milliseconds to return. Not good!